src/chainstate/stacks/index/proofs.rs

```diff
@@ -0,0 +1,1129 @@
+        let (node, _) = storage.read_nodetype(ptr)?;
+        let all_hashes = Trie::get_children_hashes(storage, &node)?;
+ 
+        let hashes = 
+            if node.is_leaf() {
+                vec![]
+            }
+            else {
+                TrieMerkleProof::make_proof_hashes(&node, &all_hashes, prev_chr)?
+            };
+    
+        let proof_node = match node {
+            TrieNodeType::Leaf(ref data) => {
+                TrieMerkleProofType::Leaf((prev_chr, data.clone()))
+            },
+            TrieNodeType::Node4(ref data) => {
+                let mut hash_slice = [TrieHash::from_data(&[]); 3];
+                hash_slice.copy_from_slice(&hashes[0..3]);
+
+                TrieMerkleProofType::Node4((prev_chr, data.clone(), hash_slice))
+            },
+            TrieNodeType::Node16(ref data) => {
+                let mut hash_slice = [TrieHash::from_data(&[]); 15];
+                hash_slice.copy_from_slice(&hashes[0..15]);
+
+                TrieMerkleProofType::Node16((prev_chr, data.clone(), hash_slice))
+            },
+            TrieNodeType::Node48(ref data) => {
+                let mut hash_slice = [TrieHash::from_data(&[]); 47];
+                hash_slice.copy_from_slice(&hashes[0..47]);
+
+                TrieMerkleProofType::Node48((prev_chr, data.clone(), hash_slice))
+            },
+            TrieNodeType::Node256(ref data) => {
+                let mut hash_slice = [TrieHash::from_data(&[]); 255];
+                hash_slice.copy_from_slice(&hashes[0..255]);
+
+                TrieMerkleProofType::Node256((prev_chr, data.clone(), hash_slice))      // ancestor hashes to be filled in later
+            }
+        };
+        Ok(proof_node)
+    }
+
+    /// Given a node's (non-backptr) ptr, and the node's backptr, make a shunt proof that links
+    /// them.  That is, make a proof that the current trie's root node hash and ptr are only reachable from the
+    /// corresponding non-backptr root in this trie's ${ptr.back_block()}th ancestor back.
+    /// s must point to the block from which we're going to walk back from.
+    /// 
+    /// The first entry of the shunt proof is the set of Trie root hashes _excluding_ the one from
+    /// backptr, as well as the index into the list of Trie root hashes into which the backptr hash
+    /// should be inserted (this root hash is calculated from the segment proof for that backptr
+    /// node).
+    ///
+    /// The last entry of the shunt proof is the set of root hashes _excluding_ the final root
+    /// hash, which will be the root hash for the segment proof for the non-backptr copy of this
+    /// node.
+    ///
+    /// All intermediate shunt proofs will contain all ancestor hashes for each node in-between the
+    /// backptr and the non-backptr node.  The intermediate root hashes will be calculated by the verifier.
+    fn make_shunt_proof(storage: &mut TrieFileStorage, backptr: &TriePtr) -> Result<Vec<TrieMerkleProofType>, Error> {
```

Comment:
It's strange that this function is called on both back pointers and pointers (in the case of the first segment proof) with two basically unrelated code paths -- it'd be better if this was made explicit:

```rust
if is_backptr {
 ... do all this ...
} else {
  ... just include the ancestor hashes ...
}
```

Reply:
Yup, will fix.

Reply:
Should be fixed now.